go/vt/sqlparser/ast.go

```diff
@@ -1725,69 +1757,184 @@ func (node OnDup) WalkSubtree(visit Visit) error {
+	return node.Lowered() == in.Lowered()
 }
 
 // EqualString performs a case-insensitive compare with str.
 func (node ColIdent) EqualString(str string) bool {
-	return cistring.CIString(node).EqualString(str)
+	return node.Lowered() == strings.ToLower(str)
+}
+
+// MarshalJSON marshals into JSON.
+func (node ColIdent) MarshalJSON() ([]byte, error) {
+	return json.Marshal(node.val)
+}
+
+// UnmarshalJSON unmarshals from JSON.
+func (node *ColIdent) UnmarshalJSON(b []byte) error {
+	var result string
+	err := json.Unmarshal(b, &result)
+	if err != nil {
+		return err
+	}
+	node.val = result
+	return nil
 }
 
 // TableIdent is a case sensitive SQL identifier. It will be escaped with
-// backquotes if it matches a keyword.
-type TableIdent string
+// backquotes if necessary.
+type TableIdent struct {
+	v string
+}
+
+// NewTableIdent creates a new TableIdent.
+func NewTableIdent(str string) TableIdent {
+	return TableIdent{v: str}
+}
 
 // Format formats the node.
 func (node TableIdent) Format(buf *TrackedBuffer) {
-	name := string(node)
-	if _, ok := keywords[strings.ToLower(name)]; ok {
-		buf.Myprintf("`%s`", name)
-		return
-	}
-	buf.Myprintf("%s", name)
+	formatID(buf, node.v, strings.ToLower(node.v))
 }
 
 // WalkSubtree walks the nodes of the subtree.
 func (node TableIdent) WalkSubtree(visit Visit) error {
 	return nil
 }
+
+// IsEmpty returns true if TabIdent is empty.
+func (node TableIdent) IsEmpty() bool {
+	return node.v == ""
+}
+
+// String returns the unescaped table name. It must
+// not be used for SQL generation. Use sqlparser.String
+// instead. The Stringer conformance is for usage
+// in templates.
+func (node TableIdent) String() string {
+	return node.v
+}
+
+// CompliantName returns a compliant id name
+// that can be used for a bind var.
+func (node TableIdent) CompliantName() string {
+	return compliantName(node.v)
+}
+
+// MarshalJSON marshals into JSON.
+func (node TableIdent) MarshalJSON() ([]byte, error) {
+	return json.Marshal(node.v)
+}
+
+// UnmarshalJSON unmarshals from JSON.
+func (node *TableIdent) UnmarshalJSON(b []byte) error {
+	var result string
+	err := json.Unmarshal(b, &result)
+	if err != nil {
+		return err
+	}
+	node.v = result
+	return nil
+}
+
+func formatID(buf *TrackedBuffer, original, lowered string) {
+	for i, c := range original {
+		if i == 0 && isDigit(uint16(c)) {
+			goto mustEscape
+		}
+		if !isLetter(uint16(c)) && !isDigit(uint16(c)) {
+			goto mustEscape
+		}
+	}
+	if _, ok := keywords[lowered]; ok {
+		goto mustEscape
+	}
+	buf.Myprintf("%s", original)
+	return
+
+mustEscape:
+	buf.WriteByte('`')
+	for _, c := range original {
+		buf.WriteRune(c)
+		if c == '`' {
+			buf.WriteByte('`')
+		}
+	}
+	buf.WriteByte('`')
+}
+
+func compliantName(in string) string {
+	var buf bytes.Buffer
+	for i, c := range in {
+		if i == 0 && isDigit(uint16(c)) {
+			buf.WriteByte('_')
+			continue
+		}
+		if !isLetter(uint16(c)) && !isDigit(uint16(c)) {
+			buf.WriteByte('_')
+			continue
+		}
```

Comment:
It's not that simple. This would work:
```
if !isLetter(uint16(c)) {
  if i == 0 && isDigit(uint16(c)) {
    buf.WriteByte('_')
    continue
  } else if !isDigit(uint16(c)) {
    buf.WriteByte('_')
    continue
  }
}

But it's less readable.